node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -1374,45 +1620,67 @@ async fn handle_collation_fetched_result<Context>(
 		},
 	};
 
-	if let Some(collations) = state.collations_per_relay_parent.get_mut(&relay_parent) {
-		if let CollationStatus::Seconded = collations.status {
-			gum::debug!(
-				target: LOG_TARGET,
-				?relay_parent,
-				"Already seconded - no longer interested in collation fetch result."
-			);
-			return
-		}
-		collations.status = CollationStatus::WaitingOnValidation;
-	}
+	let collations = &mut per_relay_parent.collations;
+	// There's always a single collation being fetched at any moment of time.
+	// In case of a failure, we reset the status back to waiting.
+	collations.status = CollationStatus::WaitingOnValidation;
 
-	if let Entry::Vacant(entry) = state.pending_candidates.entry(relay_parent) {
+	let fetched_collation = FetchedCollation::from(&candidate_receipt);
+	if let Entry::Vacant(entry) = state.fetched_candidates.entry(fetched_collation) {
 		collation_event.1.commitments_hash = Some(candidate_receipt.commitments_hash);
 
-		if let Some(pvd) = request_persisted_validation_data(
```

Comment:
I believe this is an error. `request_persisted_validation_data` is drawing the `PersistedValidationData` from the chain-state, which will yield a `PersistedValidationData` with the wrong `parent_head`.

Reply:
We should construct the `PersistedValidationData` without accessing the chain state, as this is obsolete in asynchronous backing.

Reply:
The Prospective Parachains subsystem should be able to help.

Reply:
We should also be checking that the `PersistedValidationData` that we construct matches the hash in the candidate receipt.

Reply:
The other option is to alter the collation response to include the `PersistedValidationData` or `HeadData`.